drivers: rcar_mmc: handle unaligned access to data buffer

xakep-amatop opened the pull request.

xakep-amatop committed:
drivers: rcar_mmc: handle unaligned access to data buffer

Handle unaligned access to data buffer and case when block size less
than configured width for access to SD_BUF.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: rcar_mmc: add support of 2/4 bytes width for access to SD_BUF0

Add support of 2 and 4 bytes width for access to SD_BUF0 register and
add helper functions for reading/writing from/to the register, all
necessary logic is inside helper functions. The functions return/receive
u64 type and implicitly cast it to a necessary type.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -657,11 +716,11 @@ static int rcar_mmc_sd_buf_rx_tx_data(const struct device *dev,
 
 			if (is_read) {
-				buf0 = sys_read64(DEVICE_MMIO_GET(dev) + RCAR_MMC_BUF0);
+				buf0 = rcar_mmc_read_buf0(dev);
```
Just out of curiosity, why this patches are not squashed?

xakep-amatop replied:
at the first time, they were, but I've split one commit into two: one for handling unaligned access and another for working with different sd_buf0 access width sizes

if it is needed, I can combine them with one another

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -659,11 +652,16 @@ static int rcar_mmc_sd_buf_rx_tx_data(const struct device *dev,
 
-		for (w_off = 0; (w_off * sizeof(*buf)) < data->block_size; w_off++) {
+		for (w_off = 0; w_off < aligned_block_size; w_off += sd_buf0_size) {
+			uint64_t buf0 = 0ULL;
```
I am curious what will happen with the remaining of the buffer. For example, we have only 1 block with block size = 1. Your code will send 8 bytes: 1 byte of data and 7 zeroes. Will those 7 zeroes overwrite data past the intended block?

xakep-amatop replied:
unfortunately, according to data from CSD structure (H3ULCB board), I can't use block size isn't equal to 512:
`READ_BL_LEN [83:80] == 9` and `READ_BL_PARTIAL [79] == 0` and `READ_BLK_MISALIGN [77] == 0`
`WRITE_BL_LEN [25:22] == 9` and `WRITE_BL_PARTIAL[21] == 0` and `WRITE_BLK_MISALIGN [78] == 0`

```
The data block length is computed as 2 ^ READ_BL_LEN for reading and 2 ^ WRITE_BL_LEN for writing.
```

```
WRITE_BL_PARTIAL defines whether partial block sizes can be used in block write commands.
WRITE_BL_PARTIAL=’0’ means that only the 512B and the WRITE_BL_LEN block size can
be used for block oriented data write.
WRITE_BL_PARTIAL=’1’ means that smaller blocks can be used as well. The minimum block
size is one byte.
Higher than 2 GB of density (sector access mode):
WRITE_BL_PARTIAL=’0’ means that only the 512B and the WRITE_BL_LEN block size can
be used for block oriented data write.
WRITE_BL_PARTIAL=’1’ means that smaller blocks can be used as well. The minimum block
size will be equal to minimum addressable unit, one sector (512B).
```
the same description about `READ_BL_PARTIAL`.


```
7.3.9 WRITE_BLK_MISALIGN [78]
Defines if the data block to be written by one command can be spread over more than one physical block
of the memory device. The size of the memory block is defined in WRITE_BL_LEN.
WRITE_BLK_MISALIGN=0 signals that crossing physical block boundaries is invalid.
WRITE_BLK_MISALIGN=1 signals that crossing physical block boundaries is allowed.
```

Moreover, Zephyr uses 512 block size in all places where we need to know the block size

xakep-amatop replied:
for testing the case mentioned by you, it is needed another SD or MMC with appropriate capabilities: 1-byte block writing support

lorc replied:
Okay, so if we can't test this code, how we can be sure that is working properly? Just by looking at it, I can assume that it should work fine at least for block sizes >= 8 bytes. 

I believe you need to add additional check to make sure that block size at least 8 bytes, along with a comment that we can't be sure that smaller block sizes would work at all.

xakep-amatop replied:
restore prev checks and add extra checks of block size in a new commit

xakep-amatop committed:
drivers: rcar_mmc: add extra checks for block size

MMC controller doesn't support block size above 512 and less 1.

Block size for DDR mode and multi-block xref with auto triggering of CMD12
must be fixed and equal to 512 bytes. Furthermore, in cases of multiple
block transfer without automatic issuing of CMD12, as well as 512 bytes,
32, 64, 128, and 256 bytes are specifiable.

In the case of writing of one-byte block, 8-bit width cannot be specified
for the bus width.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

lorc reviewed (commented):
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

firscity reviewed (approved):
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`
